### PR TITLE
Add Malicious Ergo Keyboard to QMK

### PR DESCRIPTION
**Pull request by awkannan**
## Description

Adding Malicious Ergo keyboard to QMK

## Types of Changes

- [ ] Core
- [ ] Bugfix
- [ ] New feature
- [ ] Enhancement/optimization
- [x] Keyboard (addition or update)
- [ ] Keymap/layout/userspace (addition or update)
- [ ] Documentation

## Issues Fixed or Closed by This PR

* 

## Checklist

- [x] My code follows the code style of this project: [**C**](https://docs.qmk.fm/#/coding_conventions_c), [**Python**](https://docs.qmk.fm/#/coding_conventions_python)
- [x] I have read the [**PR Checklist** document](https://docs.qmk.fm/#/pr_checklist) and have made the appropriate changes.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the [**CONTRIBUTING** document](https://docs.qmk.fm/#/contributing).
- [ ] I have added tests to cover my changes.
- [x] I have tested the changes and verified that they work and don't break anything (as well as I can manage).

### CODE REVIEW

File: keyboards/cannonkeys/malicious_ergo/config.h [hunk 1]
```diff
@@ -0,0 +1,94 @@
...
+
+/* Set 0 if debouncing isn't needed */
+#define DEBOUNCE    5
+
+/* Mechanical locking support. Use KC_LCAP, KC_LNUM or KC_LSCR instead in keymap */
+#define LOCKING_SUPPORT_ENABLE
+/* Locking resynchronize hack */
+#define LOCKING_RESYNC_ENABLE
+
+#define RGBLIGHT_ANIMATIONS
```
Comment (drashna) [inline keyboards/cannonkeys/malicious_ergo/config.h:56]: Could you explicitly define which animations are enabled? 
```suggestion
#define RGBLIGHT_EFFECT_BREATHING
#define RGBLIGHT_EFFECT_RAINBOW_MOOD
#define RGBLIGHT_EFFECT_RAINBOW_SWIRL
#define RGBLIGHT_EFFECT_SNAKE
#define RGBLIGHT_EFFECT_KNIGHT
#define RGBLIGHT_EFFECT_CHRISTMAS
#define RGBLIGHT_EFFECT_STATIC_GRADIENT
#define RGBLIGHT_EFFECT_RGB_TEST
#define RGBLIGHT_EFFECT_ALTERNATING
#define RGBLIGHT_EFFECT_TWINKLE
```

File: keyboards/cannonkeys/malicious_ergo/malicious_ergo.c [hunk 2]
```diff
@@ -0,0 +1 @@
+#include "malicious_ergo.h"
```
Comment (drashna) [inline keyboards/cannonkeys/malicious_ergo/malicious_ergo.c:1]: Could you add a GPL2+ compatible license header here? 

For instance:
```c
// Copyright 2021 Your Name (@yourgithub)
// SPDX-License-Identifier: GPL-2.0-or-later
```

Or 

https://github.com/qmk/qmk_firmware/blob/6b299fab6e6a3143c061b9ea005253361fcd524b/data/templates/base/keyboard.c#L1-L15

File: keyboards/cannonkeys/malicious_ergo/readme.md [hunk 3]
```diff
@@ -0,0 +1,10 @@
+# CannonKeys Malicious Ergo
+
+Keyboard Maintainer: [Andrew Kannan](https://github.com/awkannan)  
+Hardware Supported: STM32F072CBT6  
+
+Make example for this keyboard (after setting up your build environment):
```
Comment (drashna) [inline keyboards/cannonkeys/malicious_ergo/readme.md:6]: Could you add instructions on how to enter the bootloader, to flash new firmware? 

Eg: 
```markup
## Bootloader

Enter the bootloader in 3 ways:

* **Bootmagic reset**: Hold down the key at (0,0) in the matrix (usually the top left key or Escape) and plug in the keyboard
* **Physical reset button**: Briefly press the button on the back of the PCB - some may have pads you must short instead
* **Keycode in layout**: Press the key mapped to `RESET` if it is available
```

**Review by drashna — CHANGES_REQUESTED** [review]

**Comment by awkannan** [comment]
@drashna Should be all set now, thanks!

**Review by drashna — APPROVED** [review]

File: keyboards/cannonkeys/malicious_ergo/rules.mk [hunk 4]
```diff
@@ -0,0 +1,25 @@
...
+#   change yes to no to disable
+#
+BOOTMAGIC_ENABLE = yes      # Enable Bootmagic Lite
+MOUSEKEY_ENABLE = yes	# Mouse keys
+EXTRAKEY_ENABLE = yes	# Audio control and System control
+CONSOLE_ENABLE = yes	# Console for debug
+COMMAND_ENABLE = yes    # Commands for debug and configuration
+NKRO_ENABLE = yes           # Enable N-Key Rollover
+BACKLIGHT_ENABLE = yes
+RGBLIGHT_ENABLE = yes
```
Comment (fauxpark) [inline keyboards/cannonkeys/malicious_ergo/rules.mk:20]:
```suggestion
BOOTMAGIC_ENABLE = yes      # Enable Bootmagic Lite
MOUSEKEY_ENABLE = yes       # Mouse keys
EXTRAKEY_ENABLE = yes       # Audio control and System control
CONSOLE_ENABLE = yes        # Console for debug
COMMAND_ENABLE = yes        # Commands for debug and configuration
NKRO_ENABLE = yes           # Enable N-Key Rollover
BACKLIGHT_ENABLE = yes      # Enable keyboard backlight functionality
RGBLIGHT_ENABLE = yes       # Enable keyboard RGB underglow
AUDIO_ENABLE = no           # Audio output
```

File: keyboards/cannonkeys/malicious_ergo/rules.mk [hunk 5]
```diff
@@ -0,0 +1,25 @@
...
+WS2812_DRIVER = spi
+
+
```
Comment (fauxpark) [inline keyboards/cannonkeys/malicious_ergo/rules.mk:23]:
```suggestion
```

File: keyboards/cannonkeys/malicious_ergo/readme.md [hunk 6]
```diff
@@ -0,0 +1,18 @@
+# CannonKeys Malicious Ergo
+
+Keyboard Maintainer: [Andrew Kannan](https://github.com/awkannan)  
+Hardware Supported: STM32F072CBT6
```
Comment (fauxpark) [inline keyboards/cannonkeys/malicious_ergo/readme.md:4]:
```suggestion
* Keyboard Maintainer: [Andrew Kannan](https://github.com/awkannan)
* Hardware Supported: STM32F072CBT6
```

File: keyboards/cannonkeys/malicious_ergo/info.json [hunk 7]
```diff
@@ -0,0 +1,10 @@
+{
```
Comment (fauxpark) [inline keyboards/cannonkeys/malicious_ergo/info.json:1]: image

**Review by fauxpark — CHANGES_REQUESTED** [review]

File: keyboards/cannonkeys/malicious_ergo/info.json (hunk 7 above)
Comment (awkannan) [reply keyboards/cannonkeys/malicious_ergo/info.json:1]: Oh man :( These ergo layout boards are tough.

I think I used the rotated version to generate the info JSON. Should I not?

Comment (drashna) [reply keyboards/cannonkeys/malicious_ergo/info.json:1]: Nope, rotation isn't supported.

**Comment by awkannan** [comment]
@fauxpark I've finally removed rotation from the info json. Hopefully it passes now!

**Comment by awkannan** [comment]
Wanted to give this one a quick bump @fauxpark

File: keyboards/cannonkeys/malicious_ergo/info.json (hunk 7 above)
Comment (fauxpark) [reply keyboards/cannonkeys/malicious_ergo/info.json:1]: This is still outstanding.

**Review by fauxpark — CHANGES_REQUESTED** [review]

**Comment by awkannan** [comment]
@fauxpark What specifically about the info.json is still outstanding, if you don't mind me asking?

I removed rotation from the keymap layout in this commit:https://github.com/qmk/qmk_firmware/pull/17076/commits/c1dd4302e91ac0678975307dc982d86a2507404b

Is there something I am missing?

**Comment by fauxpark** [comment]
The layout is improperly aligned, and the ordering is incorrect.
image

I recommend formatting the info.json more like this:
https://github.com/qmk/qmk_firmware/blob/master/layouts/default/60_ansi/info.json
as it makes things a lot easier.

**Review by noroadsleft — CHANGES_REQUESTED** [review]
Also, are there any layout options at play here? Split vs. 2u Backspace, ANSI/ISO, anything like that?

**Comment by awkannan** [comment]
@fauxpark JSON is straight on now - it should be fixed.
@noroadsleft all keys are shown in the layout

File: keyboards/cannonkeys/malicious_ergo/info.json [hunk 8]
```diff
@@ -0,0 +1,10 @@
...
+    "keyboard_name": "Malicious Ergo",
+    "url": "https://cannonkeys.com",
+    "maintainer": "awkannan",
```
Comment (fauxpark) [inline keyboards/cannonkeys/malicious_ergo/info.json:4]:
```suggestion
    "keyboard_name": "Malicious Ergo",
    "manufacturer": "CannonKeys",
    "url": "https://cannonkeys.com",
    "maintainer": "awkannan",
    "usb": {
        "vid": "0xCA04",
        "pid": "0x0009",
        "device_version": "0.0.1"
    },
```

File: keyboards/cannonkeys/malicious_ergo/config.h [hunk 9]
```diff
@@ -0,0 +1,103 @@
...
+
+#pragma once
+
+/* USB Device descriptor parameter */
+#define VENDOR_ID       0xCA04
+#define PRODUCT_ID      0x0009
+#define DEVICE_VER      0x0001
+#define MANUFACTURER CannonKeys
+#define PRODUCT Malicious Ergo
+
```
Comment (fauxpark) [inline keyboards/cannonkeys/malicious_ergo/config.h:28]:
```suggestion
```

**Review by fauxpark — CHANGES_REQUESTED** [review]

File: keyboards/cannonkeys/malicious_ergo/info.json [hunk 10]
```diff
@@ -0,0 +1,10 @@
...
...x":16.75, "y":0}, {"x":18, "y":0}, {"x":0.25, "y":1}, {"x":1.5, "y":1, "w":1.5}, {"x":3, "y":1}, {"x":4, "y":1}, {"x":5, "y":1}, {"x":6, "y":1}, {"x":7, "y":1}, {"x":9.5, "y":1}, {"x":10.5, "y":1}, {"x":11.5, "y":1}, {"x":12.5, "y":1}, {"x":13.5, "y":1}, {"x":14.5, "y":1}, {"x":15.5, "y":1}, {"x":16.5, "y":1, "w":1.5}, {"x":18.25, "y":1}, {"x":0, "y":2}, {"x":1.3800000000000001, "y":2, "w":1.75}, {"x":3.125, "y":2}, {"x":4.13, "y":2}, {"x":5.13, "y":2}, {"x":6.13, "y":2}, {"x":7.13, "y":2}, {"x":9.88, "y":2}, {"x":10.88, "y":2}, {"x":11.88, "y":2}, {"x":12.88, "y":2}, {"x":13.88, "y":2}, {"x":14.88, "y":2}, {"x":15.88, "y":2, "w":2.25}, {"x":18.5, "y":2}, {"x":1.25, "y":3, "w":2.25}, {"x":3.5, "y":3}, {"x":4.5, "y":3}, {"x":5.5, "y":3}, {"x":6.5, "y":3}, {"x":7.5, "y":3}, {"x":9.5, "y":3}, {"x":10.5, "y":3}, {"x":11.5, "y":3}, {"x":12.5, "y":3}, {"x":13.5, "y":3}, {"x":14.5, "y":3}, {"x":15.5, "y":3, "w":1.75}, {"x":17.25, "y":3}, {"x":1.5, "y":4, "w":1.5}, {"x":4.25, "y":4, "w":1.5}, {"x":5.75, "y":4, "w":2}, {"x":7.75, "y":4, "w":1.25}, {"x":9.5, "y":4, "w":2.75}, {"x":12.25, "y":4, "w":1.5}, {"x":14.25, "y":4, "w":1.5}, {"x":16.25, "y":4}, {"x":17.25, "y":4}, {"x":18.25, "y":4}]
```
Comment (noroadsleft) [inline keyboards/cannonkeys/malicious_ergo/info.json:7]: Fixes a rounding error (they bug me).

```suggestion
            "layout": [{"x":0.5, "y":0}, {"x":1.75, "y":0}, {"x":2.75, "y":0}, {"x":3.75, "y":0}, {"x":4.75, "y":0}, {"x":5.75, "y":0}, {"x":6.75, "y":0}, {"x":7.75, "y":0}, {"x":9.75, "y":0}, {"x":10.75, "y":0}, {"x":11.75, "y":0}, {"x":12.75, "y":0}, {"x":13.75, "y":0}, {"x":14.75, "y":0}, {"x":15.75, "y":0}, {"x":16.75, "y":0}, {"x":18, "y":0}, {"x":0.25, "y":1}, {"x":1.5, "y":1, "w":1.5}, {"x":3, "y":1}, {"x":4, "y":1}, {"x":5, "y":1}, {"x":6, "y":1}, {"x":7, "y":1}, {"x":9.5, "y":1}, {"x":10.5, "y":1}, {"x":11.5, "y":1}, {"x":12.5, "y":1}, {"x":13.5, "y":1}, {"x":14.5, "y":1}, {"x":15.5, "y":1}, {"x":16.5, "y":1, "w":1.5}, {"x":18.25, "y":1}, {"x":0, "y":2}, {"x":1.375, "y":2, "w":1.75}, {"x":3.125, "y":2}, {"x":4.13, "y":2}, {"x":5.13, "y":2}, {"x":6.13, "y":2}, {"x":7.13, "y":2}, {"x":9.88, "y":2}, {"x":10.88, "y":2}, {"x":11.88, "y":2}, {"x":12.88, "y":2}, {"x":13.88, "y":2}, {"x":14.88, "y":2}, {"x":15.88, "y":2, "w":2.25}, {"x":18.5, "y":2}, {"x":1.25, "y":3, "w":2.25}, {"x":3.5, "y":3}, {"x":4.5, "y":3}, {"x":5.5, "y":3}, {"x":6.5, "y":3}, {"x":7.5, "y":3}, {"x":9.5, "y":3}, {"x":10.5, "y":3}, {"x":11.5, "y":3}, {"x":12.5, "y":3}, {"x":13.5, "y":3}, {"x":14.5, "y":3}, {"x":15.5, "y":3, "w":1.75}, {"x":17.25, "y":3}, {"x":1.5, "y":4, "w":1.5}, {"x":4.25, "y":4, "w":1.5}, {"x":5.75, "y":4, "w":2}, {"x":7.75, "y":4, "w":1.25}, {"x":9.5, "y":4, "w":2.75}, {"x":12.25, "y":4, "w":1.5}, {"x":14.25, "y":4, "w":1.5}, {"x":16.25, "y":4}, {"x":17.25, "y":4}, {"x":18.25, "y":4}]
```

**Review by noroadsleft — CHANGES_REQUESTED** [review]

**Comment by awkannan** [comment]
@noroadsleft @fauxpark Thanks for the suggestions - just rolled them in

**Review by noroadsleft — APPROVED** [review]

**Comment by noroadsleft** [comment]
Thanks!

**Pull request merged**